Make transparency levels a list.

grokys opened the pull request:
## What does the pull request do?

Instead of specifying a single transparency level, you can now specify a list of levels in decreasing order of preference.

## Platform Support

Platforms < win10.0.17000 are on a "best effort" basis.

|             | Windows 10/11 | Windows 8 | Windows 7 | MacOS | X11 | Android |
|-------------|---------------|-----------|-----------|-------|-----------|---------|
| None        | ✓ [1]         | ✓ [1]     | ✓        | ✓     | ✓ [1]     |         | 
| Transparent | ✓ [2]         | ✓ [2]     |           | ✓     | ✓ [2]     | ✓ [6]   |
| Blur        |               | ✓ [2]     |           |       | ✓ [2] [5] | ✓ [7]   |
| AcrylicBlur | ✓ [2] [3]     |           |           | ✓     |           |         |
| Mica        | ✓ [2] [4]     |           |           |       |           |         |

[1]: Only supported when composition is disabled
[2]: Only supported when composition is enabled
[3]: Acrylic is supported on Windows >= 10.0.15063
[4]: Mica is supported on Windows >= 10.0.22000
[5]: Blur is supported under KWin
[6]: Transparent is support on Android API level >= 30
[7]: Blur is supported on Android API level >= 31

## Questions

~~Currently, what happens when none of the requested transparency levels is available is platform-dependent. All backends other than X11 leave the current transparency level in place, X11 selects `None` or `Transparent` depending on whether the compositor is enabled. We should make this consistent, but what should the semantics be?~~ Fixed: now if no matching transparency level is found, a platform-specific default will be used and reported via `ActualTransparencyLevel`.

## Breaking changes

- `TransparencyLevelHint` is now a list
- `WindowTransparencyLevel` is no longer an enum

grokys committed:
Replace WindowTransparencyLevel enum with list of structs

The transparency level is not yet communicated to the backends though.

Co-Authored-By: Max Katz <maxkatz6@outlook.com>

grokys committed:
Implement transparency level hint list on win32.

Other backends have the relevant code commented-out still.

grokys committed:
Implement transparency level hint list on Android.

grokys committed:
Implement transparency level hint list on X11.

grokys committed:
Implement transparency level hint list on macOS.

grokys committed:
Tweaked android logic.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0035372-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/transparency-hints-list

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0035421-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

grokys committed:
Fix display of unsupported transparency levels.

Only make the window brush transparent when we requested a transparent window and were given one, otherwise we display a transparent brush for e.g. `None` because `None` is not available when using composition.

grokys committed:
Set default transparency level.

If no matching transparency level was found, use the default on all platforms.

Raise `TransparencyLevelChanged` on all platforms.

grokys committed:
Make stub impls return None transparency.

grokys committed:
Fix compile error.

Accidentally removed this line - it's unused but unrelated to this PR so putting it back.

grokys committed:
Fix compile error.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0035457-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

maxkatz6 committed:
Merge branch 'master' into feature/transparency-hints-list
